File: frontend/src/types/report.ts
export type ReportPeriod = "today" | "week" | "month"
export type FinancePeriod = "week" | "month"

export interface SalesChartPoint {
  label: string
  revenue: string
  sale_count: number
}

export interface SaleSummaryRow {
  id: number
  created_at: string
  item_count: number
  total: string
}

export interface SalesReport {
  period: ReportPeriod
  total_revenue: string
  sale_count: number
  chart: SalesChartPoint[]
  sales: SaleSummaryRow[]
}

export interface TopProductRow {
  product_id: number
  sku: string
  name: string
  quantity_sold: number
  revenue: string
}

export interface TopProductsReport {
  products: TopProductRow[]
}

export interface StockValueRow {
  product_id: number
  sku: string
  name: string
  stock: number
  cost_price: string
  line_value: string
}

export interface StockValueReport {
  total_value: string
  product_count: number
  items: StockValueRow[]
}

export interface PurchaseOrderChartPoint {
  label: string
  spend: string
  po_count: number
}

export interface PurchaseOrderSummaryRow {
  id: number
  created_at: string
  supplier_name: string | null
  status: string
  total: string
}

export interface PurchaseOrdersReport {
  period: ReportPeriod
  total_spend: string
  po_count: number
  open_po_count: number
  open_po_value: string
  chart: PurchaseOrderChartPoint[]
  purchases: PurchaseOrderSummaryRow[]
}

export interface LowStockRow {
  product_id: number
  sku: string
  name: string
  stock: number
  low_stock_threshold: number
  line_value: string
}

export interface DeadStockRow {
  product_id: number
  sku: string
  name: string
  stock: number
  line_value: string
}

export interface SellThroughRow {
  product_id: number
  sku: string
  name: string
  units_sold: number
  stock: number
  sell_through_pct: string
}

export interface InventoryPerformanceReport {
  low_stock_count: number
  dead_stock_value: string
  total_stock_value: string
  avg_turnover_ratio: string
  low_stock_items: LowStockRow[]
  dead_stock_items: DeadStockRow[]
  sell_through_items: SellThroughRow[]
  open_po_count: number
  open_po_items: PurchaseOrderSummaryRow[]
}

export interface CashFlowChartPoint {
  label: string
  inbound: string
  outbound: string
}

export interface OutstandingDocumentRow {
  id: number
  party_name: string
  total: string
  amount_paid: string
  outstanding: string
}

export interface VendorSpendRow {
  supplier_id: number | null
  supplier_name: string
  total_spend: string
  po_count: number
}

export interface FinanceSummaryReport {
  period: FinancePeriod
  payments_in: string
  payments_out: string
  ap_outstanding: string
  ar_outstanding: string
  revenue: string
  cogs_approx: string
  gross_margin_pct: string
  chart: CashFlowChartPoint[]
  ap_by_supplier: OutstandingDocumentRow[]
  po_spend_by_vendor: VendorSpendRow[]
}

export interface StatusCountRow {
  status: string
  count: number
}

export interface StageCountRow {
  stage: string
  count: number
  total_value: string
}

export interface CustomerGrowthPoint {
  label: string
  count: number
}

export interface MarketingFunnelReport {
  total_leads: number
  leads_by_status: StatusCountRow[]
  total_opportunities: number
  opportunities_by_stage: StageCountRow[]
  total_customers: number
  new_customers_30d: number
  customer_growth_chart: CustomerGrowthPoint[]
}

export interface RoleSummaryRow {
  role_name: string
  user_count: number
}

export interface AgentRunStatusRow {
  run_type: string
  status: string
  count: number
}

export interface ItOverviewReport {
  api_status: string
  db_status: string
  active_user_count: number
  role_count: number
  permission_count: number
  roles: RoleSummaryRow[]
  agent_runs: AgentRunStatusRow[]
}

export type ReportTabId =
  | "sales"
  | "top-products"
  | "purchase-orders"
  | "inventory-performance"
  | "stock-value"
  | "finance-summary"
  | "marketing-funnel"
  | "it-overview"

File: backend/app/schemas/report.py
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, Field


ReportPeriod = str  # today | week | month


class SalesChartPoint(BaseModel):
    label: str
    revenue: Decimal
    sale_count: int


class SaleSummaryRow(BaseModel):
    id: int
    created_at: datetime
    item_count: int
    total: Decimal


class SalesReportResponse(BaseModel):
    period: str
    total_revenue: Decimal
    sale_count: int
    chart: list[SalesChartPoint]
    sales: list[SaleSummaryRow]


class TopProductRow(BaseModel):
    product_id: int
    sku: str
    name: str
    quantity_sold: int
    revenue: Decimal


class TopProductsReportResponse(BaseModel):
    products: list[TopProductRow]


class StockValueRow(BaseModel):
    product_id: int
    sku: str
    name: str
    stock: int
    cost_price: Decimal
    line_value: Decimal


class StockValueReportResponse(BaseModel):
    total_value: Decimal
    product_count: int
    items: list[StockValueRow] = Field(default_factory=list)


class PurchaseOrderChartPoint(BaseModel):
    label: str
    spend: Decimal
    po_count: int


class PurchaseOrderSummaryRow(BaseModel):
    id: int
    created_at: datetime
    supplier_name: str | None
    status: str
    total: Decimal


class PurchaseOrdersReportResponse(BaseModel):
    period: str
    total_spend: Decimal
    po_count: int
    open_po_count: int
    open_po_value: Decimal
    chart: list[PurchaseOrderChartPoint]
    purchases: list[PurchaseOrderSummaryRow]


class LowStockRow(BaseModel):
    product_id: int
    sku: str
    name: str
    stock: int
    low_stock_threshold: int
    line_value: Decimal


class DeadStockRow(BaseModel):
    product_id: int
    sku: str
    name: str
    stock: int
    line_value: Decimal


class SellThroughRow(BaseModel):
    product_id: int
    sku: str
    name: str
    units_sold: int
    stock: int
    sell_through_pct: Decimal


class InventoryPerformanceReportResponse(BaseModel):
    low_stock_count: int
    dead_stock_value: Decimal
    total_stock_value: Decimal
    avg_turnover_ratio: Decimal
    low_stock_items: list[LowStockRow] = Field(default_factory=list)
    dead_stock_items: list[DeadStockRow] = Field(default_factory=list)
    sell_through_items: list[SellThroughRow] = Field(default_factory=list)
    open_po_count: int
    open_po_items: list[PurchaseOrderSummaryRow] = Field(default_factory=list)


class CashFlowChartPoint(BaseModel):
    label: str
    inbound: Decimal
    outbound: Decimal


class OutstandingDocumentRow(BaseModel):
    id: int
    party_name: str
    total: Decimal
    amount_paid: Decimal
    outstanding: Decimal


class VendorSpendRow(BaseModel):
    supplier_id: int | None
    supplier_name: str
    total_spend: Decimal
    po_count: int


class FinanceSummaryReportResponse(BaseModel):
    period: str
    payments_in: Decimal
    payments_out: Decimal
    ap_outstanding: Decimal
    ar_outstanding: Decimal
    revenue: Decimal
    cogs_approx: Decimal
    gross_margin_pct: Decimal
    chart: list[CashFlowChartPoint]
    ap_by_supplier: list[OutstandingDocumentRow] = Field(default_factory=list)
    po_spend_by_vendor: list[VendorSpendRow] = Field(default_factory=list)


class StatusCountRow(BaseModel):
    status: str
    count: int


class StageCountRow(BaseModel):
    stage: str
    count: int
    total_value: Decimal


class CustomerGrowthPoint(BaseModel):
    label: str
    count: int


class MarketingFunnelReportResponse(BaseModel):
    total_leads: int
    leads_by_status: list[StatusCountRow]
    total_opportunities: int
    opportunities_by_stage: list[StageCountRow]
    total_customers: int
    new_customers_30d: int
    customer_growth_chart: list[CustomerGrowthPoint]


class RoleSummaryRow(BaseModel):
    role_name: str
    user_count: int


class AgentRunStatusRow(BaseModel):
    run_type: str
    status: str
    count: int


class ItOverviewReportResponse(BaseModel):
    api_status: str
    db_status: str
    active_user_count: int
    role_count: int
    permission_count: int
    roles: list[RoleSummaryRow] = Field(default_factory=list)
    agent_runs: list[AgentRunStatusRow] = Field(default_factory=list)
